python/xorbits/_mars/dataframe/base/pivot.py

```diff
@@ -0,0 +1,296 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from typing import Any, Union
+
+import numpy as np
+import pandas as pd
+
+from ... import opcodes
+from ...core import OutputType
+from ...core.context import Context
+from ...core.custom_log import redirect_custom_log
+from ...core.operand import MapReduceOperand, OperandStage
+from ...serialization.serializables import AnyField, Int32Field
+from ...utils import enter_current_session
+from ..operands import DataFrameOperandMixin, DataFrameShuffleProxy
+from ..utils import build_concatenated_rows_frame, hash_dataframe_on, parse_index
+
+
+class DataFramePivot(MapReduceOperand, DataFrameOperandMixin):
+    _op_type_ = opcodes.PIVOT
+
+    columns = AnyField("columns", default=None)
+    index = AnyField("index", default=None)
+    values = AnyField("values", default=None)
+    shuffle_size = Int32Field("shuffle_size")
+
+    @classmethod
+    def execute_map(cls, ctx: Union[dict, Context], op: "DataFramePivot"):
+        chunk = op.outputs[0]
+        df = ctx[op.inputs[0].key]
+
+        filters = hash_dataframe_on(df, op.index, op.shuffle_size)
+
+        for index_idx, index_filter in enumerate(filters):
+            reducer_index = (index_idx, chunk.index[1])
+            ctx[chunk.key, reducer_index] = (
+                ctx.get_current_chunk().index,
+                df.iloc[index_filter],
+            )
+
+    @classmethod
+    def execute_reduce(cls, ctx: Union[dict, Context], op: "DataFramePivot"):
+        chunk = op.outputs[0]
+        input_idx_to_df = dict(op.iter_mapper_data(ctx))
+        row_idxes = sorted(input_idx_to_df.keys())
+
+        res = []
+        for row_idx in row_idxes:
+            row_df = input_idx_to_df.get(row_idx, None)
+            if row_df is not None:
+                res.append(row_df)
+        ctx[chunk.key] = pd.concat(res, axis=0)
+
+    @classmethod
+    @redirect_custom_log
```

Comment:
Is `@redirect_custom_log` and `@enter_current_session` necessary?

Reply:
Good catch, they are not needed.